### PR TITLE
Fix bug in perplexity guide calculations and update perplexity numbers. Fixes #22348

### LINKED ISSUE #22348: Possibly Incorrect Perplexity Calculation in Conceptual Guide
In the docs, Conceptual Guides->Perplexity, the code underneath the section titled "Example: Calculating perplexity with GPT-2 in 🤗 Transformers" might be wrong.  This is based on my understanding.  The specific line of example code:

https://github.com/huggingface/transformers/blob/e8cc02555ee7dce7213e624ab088d8d4d1952064/docs/source/en/perplexity.mdx?plain=1#L122

I believe it should be:

`neg_log_likelihood = outputs.loss * (trg_len - 1)`

This is because `outputs = model(input_ids, labels=target_ids)` calculates `trg_len - 1` losses (and then averages them), not `trg_len`.

You can see why in the model code:

https://github.com/huggingface/transformers/blob/68287689f2f0d8b7063c400230b3766987abf18d/src/transformers/models/gpt2/modeling_gpt2.py#L1100C5-L1106

Basically, the HF API states "Note that the labels are shifted inside the model".  Because of this design choice, the model can only ever calculate `n - 1` losses, since it has to shift the labels itself.  `shift_logits = lm_logits[..., :-1, :].contiguous()` throws away the last logit, which it can't use to calculate a loss because it isn't given a label for the last position.  `shift_labels = labels[..., 1:].contiguous()` throws away the useless first label.

It's an odd decision in the API and should perhaps be a separate feature request to fix.  Regardless, this bug report is about the Perplexity calculation guide.  Since the model is only calculating `trg_len - 1` losses, it should only multiply the loss by `trg_len - 1`, not `trg_len`.

I believe the last line would also be wrong for similar reasons: `ppl = torch.exp(torch.stack(nlls).sum() / end_loc)`

An alternative is to change the guide to calculate the loss itself. This would allow the full `trg_len` labels to be used correctly.

### PR DESCRIPTION
**Pull request by fpgaminer**
# What does this PR do?

This pull request fixes the bug mentioned in issue #22348 relating to the Perplexity concept guide in the English documentation.

It removes the multiplication by `trg_len` (which was incorrect), and uses `mean` outside the loop instead of sum and division.  This both fixes the bug and simplifies the code.

I've updated the comments in the example code to reflect the changes, as well as leaving a note about the sub-optimal nature of the code.

N.B. This pull request does not touch the documentation in other languages besides English.

N.B. These changes are based on my understanding of perplexity and how the Transformers library works.  I'm not an expert.

Finally, the perplexity numbers were updated based on the new code.  I ran the example on my machine to get those new numbers.  My setup:

```
PyTorch Version: 2.0.0
GPU: NVIDIA GeForce RTX 3090
CUDA Version: 11.7
CUDA Device Capability: (8, 6)
CUDA Toolkit Version: 11
Transformers Version: 4.27.2
```

Fixes #22348

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for the clarifications and simplifications!
